doc/man/66-scanctl.1.scd
66-scanctl(1)

# NAME

66-scanctl - Send signal to a _scandir_

# SYNOPSIS

66-scanctl [ *-h* ] [ *-z* ] [ *-v* _verbosity_ ] [ *-l* _live_ ] [ *-o* _owner_ ] _signal_

# DESCRIPTION

This is a safe wrapper around *s6-svscanctl*.

66-scandir sends _signal_ to an already running _scandir_ at _live_ where by
default _live_ is *%%livedir%%* by default or the resulting path provided by
the *-l* option. If _owner_ is not explicitely set with *-o* then the user of
the current process will be used instead.

# OPTIONS

*-h*
	Prints this help.

*-z*
	use color.

*-v* _verbosity_
	Increases/decreases the verbosity of the command.++
	- *1* : (Default) Only print error messages.++
	- *2* : Also print warning messages.++
	- *3* : Also print tracing messages.++
	- *4* : Also print debugging messages.

*-l* _live_
	An absolute path. Send the signal to a _scandir_ at _live_. By default is
	%%livedir%%. The default can also be changed at compile time by
	passing the *--livedir=*_live_ option to *./configure*. Should be within a
	writable filesystem - likely a RAM filesystem. See *66-scandir*(1).

*-o* _owner_
	Send the signal to a _scandir_ owned by _owner_ instead of the current
	owner of the process. _owner_ should have the correct permissions access to
	deal with the _scandir_.

# SIGNALS

Any signal accepted by *s6-svscanctl* can be passed just without the dash *-*
character. As a result if you wanted to send a *-t* signal, you need to use:
	66-scanctl t
Further a series of commands is also accepted in the same way:
	66-scanctl st
A few convenient keywords were added to avoid having to remember basic and
useful commands:

	- *reload*: equal to *-an*
	- *interrupt*: equal to *-i*
	- *quit*: equal to *-q*
	- *halt*: equal to *-0*
	- *reboot*: equal to *-6*
	- *poweroff*: equal to *-7*

# EXAMPLES

66-scanctl reload

Updates the process supervision tree to exactly match the services listed in
_scandir_. This command is strictly equal to :

s6-svscanctl -an /service


NEWS
Changelog for 66

In 0.2.5.2
----------

    - Fix build: Remove bytes.h oblibs header file
    
In 0.2.5.1
----------
    
    - Bugs fix
    - 66-tree -S options: if after_tree and tree have the same name,
    tree is considered as the very first tree to start.
    - 66-unmountall: do not umount SS_LIVE
    - Add SIGPWR control file at creation of .s6-svscan directory

In 0.2.5.0
----------
    
    - Bugs Fix: Bad memory allocation
    - 66-tree: add -S option which allow to order the start tree process
    even after it creation.
    
In 0.2.4.1
----------
    
    - Hot fix: fix html documentation

In 0.2.4.0
----------
    
    - Bugs fix: Bad memory allocation.
    
    - Enable again RB_DISABLE_CAD but don't crash if its not available
    and warn user
    
    - Add @optsdepends and @extdepends field at [main] section:
        - @optsdepends can be considere as: "enable one on this service 
        or none"
        - @extdepends can be considere as: "enable the service if it is 
        not already declared on a tree"
    
    - 66-in{tree,service}: 
        - in case of empty value the tools return None
        - add -n option: this avoids to display the name of the field.
        - add field at 66-intree:
            - start: displays the list of tree(s) started before
            - allowed: displays a list of allowed user to use the tree
            - symlinks: displays the target of tree's symlinks
        - add field at 66-inservice:
            - optsdepends: displays the service's optional dependencies
            - extdepends: displays the service's external dependencies
    
    - 66-shutdown skeleton: be safier and check if options are past
    
    - New tool:
        - 66-update: The 66-update program makes a complete transition 
        of trees and the live directory using a old 66 format 
        (the one being replaced) with the new 66 format.

In 0.2.3.2
----------
    
    - Fix bad annoucement at 66-svctl
    - Warn in case of bad key at parenthesis parsing process
    
In 0.2.3.1
----------
    
    - Bugs fix: bad allocation memory
    - Makefile: remove creation of empty directories. 66-tree will check
    all directories and create it if it missing.
    
In 0.2.3.0
----------
    
    - Adapt to oblibs 0.0.4.0
    - Remove RB_DISABLE_CAD at 66-boot by default and allow to enable
    it with -c option. Starting on virtual system like lxc crash with
    this option. Well be safe on every case.
    - Use of new function log_? familly from oblibs:
        - standardization of the exit code
        - standardization of the verbosity output
        - allow verbosity to 4 to display debug message
    - Doc typo fix        
    
In 0.2.2.2
----------
    
    - Adapt to skalibs 2.9.1.0
    - Remove insta_? deprecated function
    
In 0.2.2.1
----------
    
    - Typo fix at html documentation
    - Fix 66-intree without options
    
In 0.2.2.0
----------

    - Bugs fix: Always check the existing of the 66 heart directories
    - 66-info is now deprecated and splitted into two differents API.
    66-intree give informations about tree
    66-inservice give informations about services

In 0.2.1.2
----------

    - Hot fix: fix 66-info graph display.
    
In 0.2.1.1
----------

    - Hot fix: fix the build of the service dependencies graph.

In 0.2.1.0
----------

    - Bugs fix.
   
    - A synchronization is now made on reboot even with force option.
   
    - The field @name has no effects except for instantiated service but 
    can be omitted. The name of the frontend file is took as name. 
    In case of instantiated service, the field @name must contain the 
    complete name of the frontend service file -- Refer to the frontend 
    documentation page.
    
    - New skeleton file: rc.shutdown.final. This skeleton file will be run
    at the very end of the shutdown procedure, after all processes have 
    been killed and all filesystems have been unmounted, just before the 
    system is rebooted or the power turned off.
    This script normally remains empty.
   
    - Extra tools has removed and provided as an independent programs at
    https://framagit.org/Obarun/66-tools.git except for 66-echo and 
    66-unmountall program which are a dependent part of 66.
   
    - The account to run the s6-log program at the associate service logger
    can be set at compilation time by the --with-s6-log-user=USER options.
    The default is root. Obviously, the @runas field at the [logger] section
    overwritte it.
    This options set the account to run the uncaught-logs too. Also,
    this option can be overwritten with the '-l' option at 66-boot invocation
    and the '-L' option at 66-scandir invocation. 
    
In 0.2.0.4
----------
	
	- Bugs fix on memory allocation
	- Respect policies decision of user: 
		Remove -m option from 66-boot on init file
	- Add -d feature to 66-scandir 

In 0.2.0.3
----------

	- 66-parser: Fix write of configuration file, add -c|C features 
	
In 0.2.0.2
----------

	- Fix the written of user configuration file 
	- Fix oneshot and bundle status check
	- Add -l features to 66-tree, rewrite tree_unsupervise function
	
In 0.2.0.0
----------

	- New tools:
		- 66-boot.
		- 66-scanctl.
		- 66-shutdown.
		- 66-shutdownd.
		- 66-hpr.
		- 66-env.
		- 66-parser.
		- 66-which.
		- 66-echo.
		- 66-unmountall.
		- execl-subuidgid.
	- New @key field and change on frontend file:
		- [Logger] section accept a @depends field, @timestamp accept none
		as value, readiness notification is a default.
		- @hiercopy in [main] allow to copy any file or directory coming
		from the directory of the service.
	- 66-envfile is now deprecated, use execl-envfile in replacement.
	- 66-enable: add -F, -c, -C features.
	- 66-svctl is now asynchrone to launch services.
	- Man pages are available.
	- Environment variables limitation: maximum 100 files by directory,
	each file cannot contain more than 4096 bytes or 50 variables.

src/66/66-scanctl.c
/* 
 * 66-scanctl.c
 * 
 * Copyright (c) 2018-2020 Eric Vidal <eric@obarun.org>
 * 
 * All rights reserved.
 * 
 * This file is part of Obarun. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution.
 * This file may not be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file./
 */
 
#include <string.h>
#include <oblibs/log.h>
#include <oblibs/obgetopt.h>

#include <skalibs/buffer.h>
#include <skalibs/stralloc.h>
#include <skalibs/types.h>

#include <66/utils.h>

#define SIGSIZE 64

#define USAGE "66-scanctl [ -h ] [ -z ] [ -v verbosity ] [ -l live ] [ -o owner ] signal"

static inline void info_help (void)
{
  static char const *help =
"66-scanctl <options> signal\n"
"\n"
"options :\n"
"	-h: print this help\n" 
"	-z: use color\n"
"	-v: increase/decrease verbosity\n"
"	-l: live directory\n"
"	-o: handle scandir of owner\n"
;

 if (buffer_putsflush(buffer_1, help) < 0)
    log_dieusys(LOG_EXIT_SYS, "write to stdout") ;
}
static inline unsigned int lookup (char const *const *table, char const *signal)
{
  unsigned int i = 0 ;
  for (; table[i] ; i++) if (!strcmp(signal, table[i])) break ;
  return i ;
}
static inline unsigned int parse_signal (char const *signal)
{
  static char const *const signal_table[] =
  {
    "reload",
    "interrupt",
    "quit",
    "halt",
    "reboot",
    "poweroff",
    0
  } ;
  unsigned int i = lookup(signal_table, signal) ;
  if (!signal_table[i]) i = 7 ;
  return i ;
}

int send_signal(char const *scandir, char const *signal)
{
	unsigned int sig = 5 ;
	size_t siglen = strlen(signal) ;
	char csig[SIGSIZE + 1] ;
	sig = parse_signal(signal) ;
	if (sig < 6)
	{
		switch(sig)
		{
			case 0: csig[0] = 'a' ; 
					csig[1] = 'n' ;
					csig[2] = 0 ;
					break ;
			case 1: csig[0] = 'i' ;
					csig[1] = 0 ;
					break ;
			case 2: csig[0] = 'q' ; 
					csig[1] = 0 ;
					break ;
			case 3: csig[0] = '0' ;
					csig[1] = 0 ;
					break ;
			case 4: csig[0] = '6' ;
					csig[1] = 0 ;
					break ;
			case 5: csig[0] = '7' ;
					csig[1] = 0 ;
					break ;
			default: break ;
		}
	}
	else
	{
		memcpy(csig,signal,siglen) ;
		csig[siglen] = 0 ;
	}
	log_info("Sending -",csig," signal to scandir: ",scandir,"...") ;
	return scandir_send_signal(scandir,csig) ;
}

int main(int argc, char const *const *argv)
{
	int r ;
	uid_t owner = MYUID ;
	char const *signal ;
	stralloc scandir = STRALLOC_ZERO ;
	
	log_color = &log_color_disable ;
	
	PROG = "66-scanctl" ;
	{
		subgetopt_t l = SUBGETOPT_ZERO ;

		for (;;)
		{
			int opt = getopt_args(argc,argv, ">hv:l:o:z", &l) ;
			if (opt == -1) break ;
			if (opt == -2) log_die(LOG_EXIT_USER,"options must be set first") ;
			switch (opt)
			{
				case 'h' : info_help(); return 0 ;
				case 'v' : if (!uint0_scan(l.arg, &VERBOSITY)) log_usage(USAGE) ; break ;
				case 'l' : if (!stralloc_cats(&scandir,l.arg)) log_die_nomem("stralloc") ;
						   if (!stralloc_0(&scandir)) log_die_nomem("stralloc") ;
						   break ;
				case 'o' :
						if (MYUID) log_die(LOG_EXIT_USER, "only root can use -o options") ;
						else if (!youruid(&owner,l.arg)) log_dieusys(LOG_EXIT_SYS,"get uid of: ",l.arg) ;
						break ;
				case 'z' :	log_color = !isatty(1) ? &log_color_disable : &log_color_enable ; break ;
				default : log_usage(USAGE) ; 
			}
		}
		argc -= l.ind ; argv += l.ind ;
	}
	
	if (argc < 1) log_usage(USAGE) ;
	signal = argv[0] ;
	r = set_livedir(&scandir) ;
	if (r < 0) log_die(LOG_EXIT_USER,"live: ",scandir.s," must be an absolute path") ;
	if (!r) log_dieusys(LOG_EXIT_SYS,"set live directory") ;
	r = set_livescan(&scandir,owner) ;
	if (r < 0) log_die(LOG_EXIT_USER,"scandir: ", scandir.s, " must be an absolute path") ;
	if (!r) log_dieusys(LOG_EXIT_SYS,"set scandir directory") ;
	r = scandir_ok(scandir.s) ;
	if (!r)log_diesys(LOG_EXIT_SYS,"scandir: ",scandir.s," is not running") ;
	else if (r < 0) log_dieusys(LOG_EXIT_SYS, "check: ", scandir.s) ;
	
	if (send_signal(scandir.s,signal) <= 0) goto err ;			
	
	stralloc_free(&scandir) ;
	return 0 ;
	err:
		stralloc_free(&scandir) ;
		return 111 ;	
}
	

